source/channel_bootstrap.c

```diff
@@ -1522,18 +1807,207 @@ struct aws_socket *aws_server_bootstrap_new_socket_listener(
         goto cleanup_listener;
     }
 
-    if (aws_socket_start_accept(
+    struct aws_socket_listener_options options = {
+        .on_accept_result = s_on_server_connection_result,
+        .on_accept_result_user_data = server_connection_args,
+        .on_accept_start_result = NULL,
+        .on_accept_start_user_data = NULL,
+    };
+
+    if (aws_socket_start_accept(&server_connection_args->listener, connection_loop, options)) {
+        goto cleanup_listener;
+    }
+
+    return &server_connection_args->listener;
+
+cleanup_listener:
+
+    ; // This line just used to avoid expression error after the label
+
+    SETUP_SOCKET_SHUTDOWN_CALLBACKS(
+        bootstrap_options->bootstrap->allocator,
+        &server_connection_args->listener,
+        socket_shutdown_release_server_connection_args,
+        s_socket_shutdown_complete_release_server_connection_fn,
+        server_connection_args)
+
+    aws_socket_clean_up(&server_connection_args->listener);
+    return NULL;
+
+cleanup_server_connection_args:
+    s_server_connection_args_release(server_connection_args);
+
+    return NULL;
+}
+
+/* Called when a listener connection attempt task completes.
+ */
+static void s_on_listener_connection_established(struct aws_socket *socket, int error_code, void *user_data) {
+    struct server_connection_args *server_connection_args = user_data;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_CHANNEL_BOOTSTRAP,
+        "id=%p: listener connection on socket %p completed with error %d.",
+        (void *)server_connection_args->bootstrap,
+        (void *)socket,
+        error_code);
+
+    if (error_code) {
+
+        SETUP_SOCKET_SHUTDOWN_CALLBACKS(
+            server_connection_args->bootstrap->allocator,
             &server_connection_args->listener,
-            connection_loop,
-            s_on_server_connection_result,
-            server_connection_args)) {
+            socket_shutdown_release_server_connection_args,
+            s_socket_shutdown_complete_release_server_connection_fn,
+            server_connection_args)
+
+        aws_socket_clean_up(&server_connection_args->listener);
+    }
+
+    if (server_connection_args->setup_callback) {
+        server_connection_args->setup_callback(
+            server_connection_args->bootstrap, error_code, server_connection_args->user_data);
+    }
+
+    s_server_connection_args_release(server_connection_args);
+
+    return;
+}
+
+struct aws_socket *aws_server_bootstrap_new_socket_listener_async(
```

Comment:
Trivial: Apart from the requested_event_loop check and an acquire on the server_connection_args, this is identical to the non async function. Is there any way to combine the two and perform those two actions conditionally? We may even be able to do the acquire on the non-async and make sure we release it unless the refcount hitting 0 is important to the shutdown process of the non-async version of the function.

Reply:
Eventually, I removed the aws_server_bootstrap_new_socket_listener_async. Now we have a `setup_callback` in `aws_server_socket_channel_bootstrap_options`. If `setup_callback` is set, it will asynchronously invoke the `setup_callback` when the listener is ready for use.